core/components/header/_actions/switch-currency.ts

```diff
@@ -0,0 +1,79 @@
+'use server';
+
+import { SubmissionResult } from '@conform-to/react';
+import { parseWithZod } from '@conform-to/zod';
+import { revalidatePath, revalidateTag } from 'next/cache';
+import { getTranslations } from 'next-intl/server';
+import { z } from 'zod';
+
+import { client } from '~/client';
+import { graphql } from '~/client/graphql';
+import { TAGS } from '~/client/tags';
+import { getCartId, setCartId } from '~/lib/cart';
+import { setPreferredCurrencyCode } from '~/lib/currency';
+
+import { CurrencyCode } from '../fragment';
+import { CurrencyCodeSchema } from '../schema';
+
+const currencySwitchSchema = z.object({
+  id: CurrencyCodeSchema,
+});
+
+// Note: this results in a new cart being created in the new currency, so the cart ID will change
+const UpdateCartCurrencyMutation = graphql(`
+  mutation UpdateCartCurrency($input: UpdateCartCurrencyInput!) {
+    cart {
+      updateCartCurrency(input: $input) {
+        cart {
+          currencyCode
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+export const updateCartCurrency = async (cartId: string, currencyCode: CurrencyCode) => {
+  const result = await client.fetch({
+    document: UpdateCartCurrencyMutation,
+    variables: { input: { data: { currencyCode }, cartEntityId: cartId } },
+  });
+  const newCartId = result.data.cart.updateCartCurrency?.cart?.entityId;
+
+  if (newCartId) {
+    await setCartId(newCartId);
+  } else {
+    throw new Error('Failed to update cart currency', { cause: result });
+  }
+};
+
+export const switchCurrency = async (_prevState: SubmissionResult | null, payload: FormData) => {
```

Comment:
This action needs some error handling like we now do in https://github.com/bigcommerce/catalyst/blob/canary/core/app/%5Blocale%5D/(default)/product/%5Bslug%5D/_actions/add-to-cart.tsx#L243. LMK if you have any questions.